### PR TITLE
Override serverUrl in wasp deploy fly deploy command

### LINKED ISSUE #2233: Server URL gets reset on the client in new deployments on Fly
**Describe the bug**
When we use the wasp deploy fly deploy command, it looks like the server URL (env variable when building the client here: https://github.com/wasp-lang/wasp/blob/8d5530519bb5e99ec7d87caf3c807785f19954dc/waspc/packages/deploy/src/providers/fly/deploy/deploy.ts#L107) is being derived from the base name (https://github.com/wasp-lang/wasp/blob/8d5530519bb5e99ec7d87caf3c807785f19954dc/waspc/packages/deploy/src/providers/fly/DeploymentInfo.ts#L30). Can we override this somehow? We are using the deploy-action (https://github.com/wasp-lang/deploy-action) in our CI and our server URL is api.schooltendance.com. This gets reset to the fly.dev URL every time we 🚢.

**To Reproduce**
1. Deploy using the command `wasp deploy fly deploy` 
2. Add a custom domain to your server and add certs (https://wasp-lang.dev/docs/advanced/deployment/cli#using-a-custom-domain-for-your-app)
3. Deploy your client by building from .wasp/build/web-app 
`npm install && REACT_APP_API_URL=https://api.customDomain.com npm run build` 
`flyctl deploy --remote-only`
4. Now test that the client uses the correct URL that you set in the previous command. 
5. Run `wasp deploy fly deploy`. Now check the URL being used by the client. It will be reset to the fly.dev URL. 

**Expected behavior**
The user should have the ability to maintain the overwritten URL or at least be able to specify which URL to use. 

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**
This might involve changes to the CLI and also the deploy-action.

### PR DESCRIPTION
**Pull request by Case-E**
### Description

Fixes #2233 

### Select what type of change this PR introduces:
1. [ ] **Just code/docs improvement** (no functional change).
2. [x] **Bug fix** (non-breaking change which fixes an issue).
3. [ ] **New feature** (non-breaking change which adds functionality).
4. [ ] **Breaking change** (fix or feature that would cause existing functionality to not work as expected).

### Update Waspc ChangeLog and version if needed
If you did a bug fix, new feature, or breaking change, that affects waspc, make sure you satisfy the following:
1. [ ] I updated [ChangeLog.md](https://github.com/wasp-lang/wasp/blob/main/waspc/ChangeLog.md) with description of the change this PR introduces.
2. [ ] I bumped waspc version in [waspc.cabal](https://github.com/wasp-lang/wasp/blob/main/waspc/waspc.cabal) to reflect changes I introduced, with regards to the version of the latest wasp release, if the bump was needed.

### Update example apps if needed
If you did code changes and added a new feature or modified an existing feature, make sure you satisfy the following:
1. [ ] I updated `waspc/examples/todoApp` as needed (updated modified feature or added new feature) and manually checked it works correctly.
2. [ ] I updated `waspc/headless-test/examples/todoApp` and its e2e tests as needed (updated modified feature and its tests or added new feature and new tests for it).

### WORKAROUND for anyone who needs this before it is merged
0. Make sure that you have flyctl installed and logged in. You can run `flyctl auth whoami`
1. Run these commands from the root of your project
```shell
wasp build
```
2. Navigate to the build folder
```shell
cd .wasp/build/web-app
```
3. Copy the client config file into the build dir
```shell
cp ../../../fly-client.toml ./fly.toml
```
4. Install dependencies and build the client
```shell
npm install && REACT_APP_API_URL=https://sub.domain.com npm run build
```
5. Create the Dockerfile and add the following contents
```shell
touch Dockerfile
```
```shell
FROM pierrezemb/gostatic
CMD [ "-fallback", "index.html" ]
COPY ./build/ /srv/http/
```
6. Create the docker ignore file
```shell
touch .dockerignore
```
7. Deploy with this command. (optional: you can pass `--local-only` in place of `--remote-only` to build locally)
```shell
flyctl deploy --remote-only
```

### CODE REVIEW

**Comment by Case-E** [comment]
The tests were failing even without the changes, so not introduced by the changes in this PR.

**Comment by Case-E** [comment]
It looks like I am unable to test this locally. Not sure why the deploy command seems to be throwing this error: 

Screen Shot 2024-08-12 at 6 12 20 PM

`wasp-cli: npm: readCreateProcessWithExitCode: chdir: invalid argument (Bad file descriptor)`

**Comment by Case-E** [comment]
This one is ready for a review. I'll make changes to the deploy-action as well following this.

**Comment by Case-E** [comment]
Everything works as expected ]. will bump the version and update the changelog.

**Comment by infomiho** [comment]
Hey @Case-E, thank you for the contribution.

We'll need to do a few things before we can deploy:
- You'll need to [sync your fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) to make sure it includes the latest code
- You'll need to merge `main` into your branch 
- I'll need to review the code changes and test it out locally
- You'll need to make sure the CI is passing (probably syncing the fork will do the trick)

File: waspc/packages/deploy/src/providers/fly/deploy/deploy.ts [hunk 1]
```diff
@@ -103,8 +104,11 @@ async function deployClient(deploymentInfo: DeploymentInfo<DeployOptions>, { bui
 	copyProjectClientTomlLocally(deploymentInfo.tomlFilePaths);
 
 	waspSays('Building web client for production...');
+	waspSays('If you configured a custom domain for the server, please run the command with an env variable: REACT_APP_API_URL=https://serverUrl.com wasp deploy fly deploy');
```
Comment (infomiho) [inline waspc/packages/deploy/src/providers/fly/deploy/deploy.ts:107]:
```suggestion
	waspSays('If you configured a custom domain for the server, you should run the command with an env variable: REACT_APP_API_URL=https://serverUrl.com wasp deploy fly deploy');
```

File: waspc/packages/deploy/src/providers/fly/deploy/deploy.ts [hunk 2]
```diff
@@ -103,8 +104,11 @@ async function deployClient(deploymentInfo: DeploymentInfo<DeployOptions>, { bui
...
+
+	const serverUrl = stripTrailingSlash(process.env.REACT_APP_API_URL ?? '') || deploymentInfo.serverUrl;
```
Comment (infomiho) [inline waspc/packages/deploy/src/providers/fly/deploy/deploy.ts:109]:
```suggestion
	const serverUrl = process.env.REACT_APP_API_URL
		? stripTrailingSlash(process.env.REACT_APP_API_URL)
		: deploymentInfo.serverUrl;
```
I'd do this to make the logic more explicit, right now there is `??` and `||` which have their own logic and this makes this a bit harder to read.

File: waspc/packages/deploy/src/providers/fly/helpers/helpers.ts [hunk 3]
```diff
@@ -141,3 +141,10 @@ export async function silence(
 	$.verbose = verboseSetting;
 	return proc;
 }
+
+export function stripTrailingSlash(url: string): string {
+	if (url.endsWith('/')) {
+		return url.slice(0, -1);
+	}
+	return url;
+}
```
Comment (infomiho) [inline waspc/packages/deploy/src/providers/fly/helpers/helpers.ts:150]: We need to add a newline at the end of this file.

**Review by infomiho — COMMENTED** [review]
Tested it locally, works as expected.

Left some minor comments.

File: waspc/packages/deploy/src/providers/fly/deploy/deploy.ts (hunk 2 above)
Comment (Case-E) [reply waspc/packages/deploy/src/providers/fly/deploy/deploy.ts:109]: That makes sense, definitely more readable with the ternary.

**Comment by Case-E** [comment]
Thanks @infomiho I didn't realise there were more changes pushed upstream from when I branched out. 

I've made the changes. As this needs to go into the current release, do I need to update the version (patch) or would y'all be batching multiple fixes into a single patch? Also should I add to the changelog?

**Comment by Case-E** [comment]
Looks like the CI failed to cancel out existing running jobs
https://github.com/wasp-lang/wasp/actions/runs/10456113609/job/28952663383?pr=2234

Screenshot 2024-08-19 at 9 01 14 PM

**Comment by infomiho** [comment]
@Case-E sorry for taking so long to get back to you, thank you for your hard work on this. 

I realised now that we can remove the use of `stripTrailingSlash` since the client does that here: https://github.com/wasp-lang/wasp/blob/main/waspc/data/Generator/templates/sdk/wasp/client/config.ts#L4

So, remove that helper and the use of it and we'll merge it! I'll update the Changelog with this change later.

**Review by infomiho — APPROVED** [review]

**Comment by Case-E** [comment]
When you get a chance, could you also do an initial review of the deploy action — https://github.com/wasp-lang/deploy-action/pull/2? I haven't tested that yet though, will do it after this change is merged and released.

**Pull request merged**